### ISSUE TITLE
Support exporting and loading ONNX models

### ISSUE THREAD

**Issue by sammlapp** [kitzeslab/opensoundscape#556]
- [ ] We should be able to export a model to ONNX so that someone could run predictions using the model without opensoundscape.

- [ ] Second, we should be able to load an ONNX model and generate predictions. If this is best done with a simple torch script rather than by implementing something in opensoundscape, that's fine - we can just add documentation of how to do this

- [ ] Third, we should be able to load and ONNX model into opensoundscape such that we could re-train (eg, warm-starting) within opensoundscape

It may be necessary or at least logical to use torchaudio to incorporate preprocessing steps into the model, as mentioned in #337 

Be aware of the numpy & built-in types [caveats](https://pytorch.org/docs/stable/onnx.html#avoid-numpy-and-built-in-python-types) for the torch.onnx module

**Comment by sammlapp** [comment]
Because ONNX will require a numeric vector input, the input will logically be either (a) audio samples or (b) a pre-processed 2d representation of the audio such as a spectrogram, potentially with multiple channels. The advantage of passing the audio sample vector (wav) is that all preprocessing parameters will be included in the model; the only thing the user has to get right is the audio sampling rate. Option (b) gives more flexibility because pre-processing does not need to be packaged into the model, but allows more opportunity for pre-processing operations and parameters to be lost or mis-implemented when the model changes hands.

**Comment by sammlapp** [comment]
Pytorch support for stft with ONNX is still a [work in progress](https://github.com/pytorch/pytorch/issues/81075)

**Comment by sammlapp** [comment]
Apparently "torch.onnx.dynamo_export" will add some onnx operators. Also, we could apparently do some custom handling to use implemented onnx functions but I don't fully understand how (see https://github.com/Alexey-Kamenev/tensorrt-dft-plugins/blob/main/tests/test_dft.py#L35 and https://github.com/pytorch/pytorch/issues/81075#issuecomment-1530713416)

Modulus has done something similar https://github.com/NVIDIA/modulus/blob/main/modulus/models/afno/afno.py#L140 

If/when we implement something like this, we will need all preprocessing steps (for inference in the exported ONNX model) to be _part of_ the pytorch model, ie layers with forward methods. This raises the question of whether we will end up entirely changing from the use of `librosa` and `scipy` to directly using the `torchaudio` API.

**Comment by sammlapp** [comment]
apparently now should work!

**Comment by sammlapp** [comment]
there's a new torch issue to follow for fft export to ONNX: https://github.com/pytorch/pytorch/issues/113067

**Comment by sammlapp** [comment]
stft + onnx seems to still not be ready (https://github.com/pytorch/pytorch/issues/113067#issuecomment-1892530038)

**Comment by sammlapp** [comment]
not sure how stable this library is, but it provides stft, spectrogram, and melspectrogram exportable to ONNX and CoreML
https://github.com/adobe-research/convmelspec/tree/main

**Comment by sammlapp** [comment]
now follow this: https://github.com/pytorch/pytorch/issues/135087

**Comment by sammlapp** [comment]
https://github.com/pytorch/pytorch/issues/112844

**Comment by sammlapp** [comment]
The current status is: we can export and use an onnx model, but ony with a fixed batch size. It seems that "dynamic shapes" should be possible to fix, but is not yet fixed
https://github.com/pytorch/pytorch/issues/113067#issuecomment-2692315050

**Comment by sammlapp** [comment]
aaand there's a new issue to follow:
https://github.com/pytorch/pytorch/issues/148687#event-16622891450

**Comment by sammlapp** [comment]
These issues are all finally resolved! This works:

save an onnx file with a Spectrogram transform

```python
import torch
import torchvision
class SpecMaker(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.transforms = torchvision.transforms.Compose(
            [
                torchaudio.transforms.Spectrogram(
                    n_fft=512,
                    win_length=512,
                    hop_length=256,
                ),
                torchaudio.transforms.AmplitudeToDB(top_db=100),
            ]
        )

    def forward(self, x):
        return self.transforms(x)


specmodel = SpecMaker()
input = torch.rand(32000 * 10)  # opso.birds
spec = specmodel(input)
input_batch = torch.stack([input, input])
spec_batch = specmodel(input_batch)

onnx_program = torch.onnx.export(
    specmodel,
    (input_batch,),
    dynamic_shapes=[{0: "dim_x"}],
    report=True,
    dynamo=True,
)

onnx_program.save("specmodel_dynamic.onnx")
```

load the onnx program and use variable batch sizes
```python
import onnx, onnxruntime

onnx_model = onnx.load("specmodel_dynamic.onnx")
onnx.checker.check_model(onnx_model)

input = torch.rand(32000 * 10)
input_batched = torch.stack([input, input, input])

EP_list = ["CUDAExecutionProvider", "CPUExecutionProvider"]

ort_session = onnxruntime.InferenceSession("specmodel_dynamic.onnx", providers=EP_list)

def to_numpy(tensor):
    return (
        tensor.detach().cpu().numpy() if tensor.requires_grad else tensor.cpu().numpy()
    )


# compute ONNX Runtime output prediction
ort_inputs = {ort_session.get_inputs()[0].name: to_numpy(input_batched)}
ort_outs = ort_session.run(None, ort_inputs)
```

**Comment by sammlapp** [comment]
well, I got a prototype working for combining torchaudio preprocessing into a model for onnx export, but the PyTorch cnn architectures are causing errors on onnx export (trying with resnet18)

**Comment by sammlapp** [comment]
Ok it works in pytorch 2.8.0, just not in the current nightly pytorch build. I implemented onnx saving of opso.CNN models in the `onnx` branch.
